app/gui2/src/components/GraphEditor/snap.ts

```diff
@@ -0,0 +1,152 @@
+import { partitionPoint } from '@/util/array'
+import type { Opt } from '@/util/opt'
+import { Rect } from '@/util/rect'
+import { Vec2 } from '@/util/vec2'
+import { abs } from 'lib0/math'
+import type { ExprId } from 'shared/yjsModel'
+import { computed, markRaw, type ComputedRef } from 'vue'
+
+export class SnapGrid {
+  leftAxes: number[]
+  rightAxes: number[]
+  topAxes: number[]
+  bottomAxes: number[]
+
+  constructor(rects: Rect[]) {
+    markRaw(this)
+    this.leftAxes = Array.from(rects, (rect) => rect.left).sort((a, b) => a - b)
+    this.rightAxes = Array.from(rects, (rect) => rect.right).sort((a, b) => a - b)
+    this.topAxes = Array.from(rects, (rect) => rect.top).sort((a, b) => a - b)
+    this.bottomAxes = Array.from(rects, (rect) => rect.bottom).sort((a, b) => a - b)
+    console.log(this.leftAxes)
+    console.log(this.rightAxes)
+    console.log(this.topAxes)
+    console.log(this.bottomAxes)
+  }
+
+  snappedMany(rects: Rect[], threshold: number): Vec2 {
+    const minSnap = rects.reduce<[number | null, number | null]>(
```

Comment:
you can construct a `Vec2` immediately and mutate it - i think it might be clearer than using tuples. alternatively using an object is also a solid alternative

as an added bonus you'd be able to avoid the function call overhead of `.reduce`'s callback... at the cost of creating an iterator for a `for...of...`. but the cost of iterators is something we can worry about some other day

Reply:
Vec2 does not allow putting `nulls`.

Reply:
ah, minSnap allows null